### PR TITLE
Truncate GOLD on-policy prompts before generation, keeping the prompt end

### PR DESCRIPTION
**Pull request by roycho96**
# What does this PR do?

GOLD generates on-policy completions from the full prompt, but then cuts the prompt from the **start** before saving the training example. So the model generates from one prompt and the student is trained on a different one.

When `max_length` is set and the prompt is longer than `max_length - max_new_tokens`, the part that is kept is the beginning of the prompt, which drops the actual question and the assistant marker at the end.

## Repro

vLLM is faked here since the cut happens in trl, not in vLLM.

```python
from types import SimpleNamespace
import torch
from transformers import AutoTokenizer
from trl.experimental.gold.gold_trainer import GOLDTrainer

tok = AutoTokenizer.from_pretrained("HuggingFaceTB/SmolLM2-360M-Instruct")
text = tok.apply_chat_template(
    [{"role": "user", "content": "What is 2 + 2? Answer with one number only."}],
    tokenize=False, add_generation_prompt=True,
)
prompt_ids = tok(text, return_tensors="pt")["input_ids"]  # ends with the assistant marker

class RecordingVLLM:
    prompts = None
    def sync_weights(self): pass
    def generate(self, prompts, images, num_generations):
        self.prompts = prompts  # what the model actually generates from
        return None, [[42]], None, None

trainer = GOLDTrainer.__new__(GOLDTrainer)
trainer.accelerator = SimpleNamespace(device=torch.device("cpu"), is_main_process=True)
trainer.processing_class = tok
trainer.args = SimpleNamespace(max_length=10, report_to=[])  # budget = 10 - 4 = 6
trainer.use_vllm = True
trainer.use_uld_loss = False
trainer.teacher_tokenizer = None
trainer.uld_loss_fn = None
trainer.vllm_generation = RecordingVLLM()
trainer.vllm_sync_frequency = 1
trainer._last_vllm_sync_step = -1
trainer.state = SimpleNamespace(global_step=0)
trainer.num_generations = 1
trainer.generation_config = SimpleNamespace(max_new_tokens=4)
trainer._buffered_inputs = [None]
trainer._buffered_text_logs = [None]

slices = [{"prompts": prompt_ids, "prompt_attention_mask": torch.ones_like(prompt_ids)}]
GOLDTrainer._generate_on_policy_for_slices(trainer, slices, [0])

generated_on = tok.decode(trainer.vllm_generation.prompts[0])
trained_on = trainer._buffered_inputs[0]["original_prompt_text"][0]
print("generated on:", repr(generated_on))
print("trained on  :", repr(trained_on))
print("same prompt for generation and training:", generated_on == trained_on)
print("training prompt keeps the assistant marker:", "assistant" in trained_on)
```

On `main`:

```
generated on: '<|im_start|>system\nYou are a helpful AI assistant named SmolLM, trained by Hugging Face<|im_end|>\n<|im_start|>user\nWhat is 2 + 2? Answer with one number only.<|im_end|>\n<|im_start|>assistant\n'
trained on  : '<|im_start|>system\nYou are a'
same prompt for generation and training: False
training prompt keeps the assistant marker: False
```

The model sees the real question, but the student is trained on just the first 6 tokens of the system prompt, with no question and no marker.

With this PR:

```
generated on: '<|im_end|>\n<|im_start|>assistant\n'
trained on  : '<|im_end|>\n<|im_start|>assistant\n'
same prompt for generation and training: True
training prompt keeps the assistant marker: True
```

## Fix

Cut the prompt to `max_length - max_new_tokens` before generation, keeping the **end** (where the assistant marker is), in both the vLLM and non-vLLM paths. This way the model generates from and is trained on the same prompt. The old cut in `_process_completions_to_buffer` is removed, and the shortened `prompts` / `prompt_attention_mask` are written back so the padding width stays right.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.

---

> [!NOTE]
> **Medium Risk**
> Changes on-policy prompt shaping for GOLD training (vLLM and `model.generate`), which directly affects what context the student sees and what the teacher re-encodes; behavior shifts for long prompts under `max_length`, though covered by new tests.
> 
> **Overview**
> Fixes a **train/generate mismatch** in GOLD on-policy training when `max_length` is set: prompts are now capped to `max_length - max_new_tokens` **before** generation, keeping the **end** of the prompt (user question + assistant marker) instead of truncating from the start only when buffering.
> 
> **vLLM path:** `_generate_on_policy_for_slices` applies keep-end budgeting on mask-stripped prompt token lists before `vllm_generation.generate`.
> 
> **Non-vLLM path:** `_generate_non_vllm_for_slices` mirrors that logic on **real** (unmasked) token counts and only rebuilds padded `prompts` / `prompt_attention_mask` when a row overflows—so left-padded rows that already fit are not wrongly trimmed.
> 
> **Buffering:** `_process_completions_to_buffer` no longer re-truncates prompts (removed `truncation_side` handling); it pads pre-budgeted IDs and **writes back** `prompts` and `prompt_attention_mask` so `prompt_length` stays consistent for the teacher.
> 
> Tests are updated/added for vLLM keep-end budgeting, non-vLLM parity, and the padded-width edge case.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 87476eaa617994d10364471720cba10dbdc9deae. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/experimental/test_gold_trainer.py [hunk 1]
```diff
@@ -646,10 +643,134 @@ def decode(self, ids, skip_special_tokens=False, clean_up_tokenization_spaces=Fa
...
+    assert torch.equal(buffered_inputs["prompt_attention_mask"], torch.tensor([[1, 1]], dtype=torch.long))
+
+
+def test_non_vllm_on_policy_budgets_prompt_before_generation(monkeypatch):
+    """Non-vLLM path budgets the prompt to max_length - max_new_tokens (keep-end) before model.generate,
+    matching the vLLM path, so the student generates from and trains on the same in-budget context."""
+
+    class DummyModel:
+        def generate(self, input_ids, attention_mask, generation_config, return_dict_in_generate):
+            # Edit 5: generate must receive the budgeted prompt (last 2 columns), not the full width-4 prompt.
```
Comment (cmpatino) [inline tests/experimental/test_gold_trainer.py]: Remove the comments that appear to be from an AI workflow.

File: tests/experimental/test_gold_trainer.py [hunk 2]
```diff
@@ -646,10 +643,134 @@ def decode(self, ids, skip_special_tokens=False, clean_up_tokenization_spaces=Fa
...
     GOLDTrainer._generate_on_policy_for_slices(trainer, slices, [0])
 
+    # prompt_max_length = max_length - max_new_tokens = 3 - 1 = 2; budgeted BEFORE generation, keeping the
+    # END of the prompt (the generation marker), so [5, 13, 6] -> [13, 6] and the student generates from it.
+    assert trainer.vllm_generation.prompts == [[13, 6]]
+
+    buffered_inputs = trainer._buffered_inputs[0]
+    assert torch.equal(buffered_inputs["input_ids"], torch.tensor([[13, 6, 42]], dtype=torch.long))
+    assert buffered_inputs["original_prompt_text"] == ["<special> B"]
+    # F8: prompts/prompt_attention_mask written back at the new padding width
```
Comment (cmpatino) [inline tests/experimental/test_gold_trainer.py]: Remove the comments that appear to be from an AI workflow

**Review by cmpatino — COMMENTED** [review]
Thank you for this change! It looks good to me, but I'm not approving it yet because of two nits about comments left by (what seems to be) an AI agent.

Also, the code quality check is failing. This is usually solved by running `make precommit` as specified in the contributing [guide for contributing](https://github.com/huggingface/trl?tab=contributing-ov-file#submitting-a-pull-request-pr).

File: tests/experimental/test_gold_trainer.py (hunk 1 above)
Comment (roycho96) [reply tests/experimental/test_gold_trainer.py]: Fixed

File: tests/experimental/test_gold_trainer.py (hunk 2 above)
Comment (roycho96) [reply tests/experimental/test_gold_trainer.py]: Removed

**Comment by roycho96** [comment]
@cmpatino Thanks for the review. Removed both comments and ran make precommit.

**Review by kashif — APPROVED** [review]
